File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <bits/stdc++.h>

using namespace std;

int n, m;
vector<int> trees;
long long mn = INT_MAX;
long long mx;

long long count(long long h) {
    long long cnt = 0;
    for (int t: trees) {
        if (h <= t)
            cnt += t - h;
    }
    return cnt;
}

void init() {
    cin >> n >> m;
    while (n--) {
        long long w;
        cin >> w;
        trees.push_back(w);
        mn = min(mn, w);
        mx = max(mx, w);
    }
}

void solve() {
    long long s = 0;
    long long e = mx + 1;
    long long mid;
    long long ans = 0;
    while (s < e) {
        mid = (s + e) / 2;
        if (s == mid || e == mid) {
            ans = mid;
            break;
        }
        long long cnt = count(mid);
        if (cnt == m)
            ans = max(mid, ans);

        if (cnt >= m) s = mid;
        else e = mid;
    }
    cout << ans;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);


    init();
    solve();
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    public static BufferedReader br;
    public static int n, m;
    public static List<Integer> trees ;
    public static long mn = Long.MAX_VALUE;
    public static long mx;

    public static long count(long h) {
        long cnt = 0;
        for (int t : trees) {
            if (h <= t)
                cnt += (t - h);
        }
        return cnt;
    }

    public static void init() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        trees = trees = new ArrayList<>();
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        while (n-- > 0) {
            int w = Integer.parseInt(st.nextToken());
            trees.add(w);
            mn = Math.min(mn, w);
            mx = Math.max(mx, w);
        }
    }

    public static void solve() {
        long s = 0;
        long e = mx + 1;
        long mid;
        long ans = 0;

        while (s < e) {
            mid = (s + e) / 2;
            if (s == mid || e == mid) {
                ans = mid;
                break;
            }
            long cnt = count(mid);
            if (cnt == m)
                ans = Math.max(mid, ans);

            if (cnt >= m) s = mid;
            else e = mid;
        }
        System.out.println(ans);
    }

    public static void main(String[] args) throws Exception {
        init();
        solve();
    }
}
